### PR TITLE
Fix httpclient gzip decoding failed #1189 - gzip header check fix

### PR DESCRIPTION
**Pull request by pecanw**
**Subsystem**
ktor-utils

**Motivation**
Fix httpclient gzip decoding failed #1189

**Solution**
Changing the gzip header check to conform [specification](http://www.zlib.org/rfc-gzip.html)

### CODE REVIEW

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt [hunk 1]
```diff
@@ -48,11 +48,32 @@ private fun CoroutineScope.inflate(
         val header = source.readPacket(GZIP_HEADER_SIZE)
         val magic = header.readShortLittleEndian()
         val format = header.readByte()
-        val padding = header.readBytes()
+        val flags = header.readByte().toInt()
+        // val time = header.readInt()
+        // val extraFlags = header.readByte()
+        // val osType = header.readByte()
```
Comment (Nohus) [inline ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: Why are the commented out variables necessary? May be worth either deleting them or using them.

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt [hunk 2]
```diff
@@ -48,11 +48,32 @@ private fun CoroutineScope.inflate(
...
+            bit 2   FEXTRA
+            bit 3   FNAME
+            bit 4   FCOMMENT
+            bit 5   reserved
+            bit 6   reserved
+            bit 7   reserved
+        */
+        if (flags and 0b0000_0100 != 0) { //FLG.FEXTRA
+            val extraLen = source.readShort().toInt()
+            val extra = source.readPacket(extraLen)
```
Comment (Nohus) [inline ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: I suspect this is here only to skip past the extra packet, a comment would help make this visible.

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt (hunk 1 above)
Comment (pecanw) [reply ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: I added an extra comment, but I would keep them for reference or future usage.

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt (hunk 2 above)
Comment (pecanw) [reply ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: I did some refactoring to make the code even more readable and added a comment as proposed.

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt [hunk 3]
```diff
@@ -13,6 +13,17 @@ import java.util.zip.*
 
 private const val GZIP_HEADER_SIZE: Int = 10
 
+// GZIP header flags bits
+private object FLG {
```
Comment (cy6erGn0m) [inline ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: GzipHeaderFlags ?

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt [hunk 4]
```diff
@@ -48,11 +59,31 @@ private fun CoroutineScope.inflate(
         val header = source.readPacket(GZIP_HEADER_SIZE)
         val magic = header.readShortLittleEndian()
         val format = header.readByte()
-        val padding = header.readBytes()
+        val flags = header.readByte().toInt()
```
Comment (cy6erGn0m) [inline ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: Memory leak here: header should be consumed entirely or discarded

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt (hunk 3 above)
Comment (pecanw) [reply ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: The `FLG` name was taken from the RFC spec, but I agree that `GzipHeaderFlags` is better. Renamed in commit e209b0b90101794b3505731012c5c0cb1d5d26b1.

File: ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt (hunk 4 above)
Comment (pecanw) [reply ktor-utils/jvm/src/io/ktor/util/EncodersJvm.kt]: Fixed in 6bdf82850036b0c532307429bc460c988263dddd.

**Comment by pecanw** [comment]
> LGTM, but we should fix memory leak(by consuming) all header bytes.

All issues from comments have been fixed, so I believe you can merge the code. @cy6erGn0m & @Nohus - can you please close your comments?

**Comment by pecanw** [comment]
However some unit test (not related to the changed code) have failed with a timeout. I haven't figured out how to re-run the build. Can someone check it please?

**Comment by pecanw** [comment]
When are you going to release a new version (1.3.2?) including this fix?

**Comment by e5l** [comment]
Yep, but no dates for now. Probably on the next week.

**Comment by pecanw** [comment]
> Yep, but no dates for now. Probably on the next week.

Any news? Can you release a new version please?

**Pull request merged**